Add TypeScript support to React Native

LinusU opened the pull request:
**Summary**

This PR adds support for TypeScript out-of-the-box. Existing codebases should work as before, but React Native should be able to pick up `.ts` and `.tsx` files automatically, and transpile them using Babel.

Note that no actual type-checking will be done, this is in line with how the current Flow support is working. The user can run `tsc --noEmit` to type-check the project.

**Test plan**

I'm currently working on this. I haven't been able to figure out how to test these changes in a local RN application.

I'll be meeting up with @kelset today, hopefully he can help me with this 💌 

With these changes, the following RN project should be able to work without further configuration: https://github.com/LinusU/react-native-babel-typescript-test

LinusU commented:
Yeah, looked at that as well, and inspecting that file I can only find one declaration of `restart` 🤔 

Won't have much more time during the day to look at this though...

rafeca commented:
> TypeError: /home/circleci/metro/packages/metro/src/commands/serve.js: Duplicate declaration "restart" (This is an error on an internal node. Probably an internal error)

I have a commit that's going to land soon that fixes this error

LinusU commented:
@rafeca nice! I worked around it in a really strange way:  https://github.com/facebook/metro/pull/209/commits/4e0048842af6059b102fecf515e9d5488cbcbca5

Seems like it's a bug inside of Babel?

rafeca commented:
Yes, it's a bug in babel scope counting ([more info](https://github.com/babel/babel/issues/8387)) that causes an issue in our inline requires transformer.

I've worked around the issue by disabling the inline requires plugin in our tests (that's fair because we don't use inline requires to transpile Metro)

LinusU commented:
> I've worked around the issue by disabling the inline requires plugin in our tests (that's fair because we don't use inline requires to transpile Metro)

Awesome!

I've rebased on `master` and the problem went away.

I also changed the approach to be a bit simpler, now the only thing added is an override for `.ts` and `.tsx` files. This approach was made possible in v7.0.0-beta.48 of Babel, and we are using a newer version then that.

I've opted to enable JSX in both `.ts` and `.tsx` files, since it's enabled for `.js` files and doesn't require the `.jsx` file ending. I'd be happy to change this if having it enabled only for `.tsx` files would be preferable.

LinusU commented:
If anyone is interested, the size of the two new dependencies are:

```text
plugin-syntax-typescript-7.0.0-beta.54.tgz          1053 bytes
plugin-transform-typescript-7.0.0-beta.54.tgz       3896 bytes
```

So adding TypeScript support to React Native will only add 4,949 bytes of extra download 😄

LinusU commented on packages/metro-react-native-babel-preset/src/configs/main.js:
```diff
@@ -124,6 +128,14 @@ const getPreset = (src, options) => {
     plugins: defaultPlugins.concat(extraPlugins),
+    overrides: [
+      {
+        test: file => isTypeScriptSource(file || 'internal.js'),
```
Sure thing!

rafeca commented on packages/metro/src/integration_tests/__tests__/__snapshots__/basic_bundle-test.js.snap:
```diff
@@ -194,6 +197,16 @@ __d(function (global, _$$_REQUIRE, module, exports, _dependencyMap) {
+  var type = 'TypeScript';
+  exports.type = type;
+  var test = true;
+  exports.test = test;
```
nice!!! 👍

LinusU commented:
Thank you for landing this 🚀

LinusU commented:
@rafeca would it be possible to cut a release, `0.43.3`, so that I can easier test this out? Thanks!

rafeca commented:
yes, I was planning to release `0.43.3` today! 🏃

LinusU commented:
Awesome 👏 thanks!

emin93 commented:
Wow this is really great thanks a lot! I just added a note to the TypeScript template (https://github.com/emin93/react-native-template-typescript) that I will stop maintaining it as it will be supported out of the box in future React Native versions. 😄

nicolashemonic commented:
Thanks! Love it!
